### PR TITLE
Add ability for crew to be killed on attack.

### PR DESCRIPTION
**Pull request by Nechochwen-D**
Add "hull breaching" weapon attribute.
Change "piercing" weapon attribute to "shield piercing" for clarity while mantaining backwards compatibility.
Add "radiation" weapon attribute.
Add "toxin" weapon attribute. (currently un-used)
Provoke on crew death.

### CODE REVIEW

**Comment by Nechochwen-D** [comment]
Currently looking for feedback on how well the values work for people. I do not currently have any outfits using the attributes but could add some if desired.

**Comment by MessyMix** [comment]
Original thread for reference: #1503 

So... after months of work, have you finally implemented it? And successfully, too? I personally think it's a little early to implement into base-game weaponry, although adding the mechanic itself doesn't seem to be a bad idea, provided that it is balanced and defect-free.

More illegal weaponry, just like nukes, would certainly fit into the current Syndicate/extremist repertoire (totally not working on the Syndicate arc rn~). I could entirely find a way to integrate the mechanic into the story, provided that mortiferous weaponry (isn't that such a nice alternative I found to "crew attack weapons"? Maybe you should consider naming these types of weapons by this alternative) are actually merged into the base game.

**Comment by Nechochwen-D** [comment]
I actually drifted away from the game for a while and just got back to it. I have tested it to a degree and everything seems to work as it is supposed to but I do not have time to properly playtest it.

I do have several other things to do from the QOL and balance perspectives but figured I would see what people think of the basic idea first.

File: source/Ship.cpp [hunk 1]
```diff
@@ -1992,6 +1993,46 @@ int Ship::TakeDamage(const Projectile &projectile, bool isBlast)
 			ApplyForce((hitForce / distance) * d);
 	}
 	
+	double hullBreaching = weapon.HullBreaching();
+	double radiation = weapon.Radiation();
+	double toxin = weapon.Toxin();
+	
+	double bunks = attributes.Get("bunks");
+        double shipSize = attributes.Get("ship size");
```
Comment (toilethinges) [inline source/Ship.cpp]: Strange spacing here. Is that on purpose?

Comment (Nechochwen-D) [inline source/Ship.cpp]: That was accidental. I will fix it upon a future push.

**Comment by endless-sky** [comment]
I'm worried that there's basically no way to balance this: the moment you have a weapon that can kill crew without boarding, it becomes possible to capture any ship you can disable.

**Comment by DingusShingleton** [comment]
It would be a major help if the crew killing weapon killed a percentage of crew. That way it would be impossible to kill a ship down to 0 crew. Make the percentage really low and you might be able to trigger some minor "Not enough crew" errors with it, but diminishing returns would make it a pretty limited strategy.

**Comment by BorkBorkGoesTheCode** [comment]
>I'm worried that there's basically no way to balance this: the moment you have a weapon that can kill crew without boarding, it becomes possible to capture any ship you can disable.

https://en.wikipedia.org/wiki/Mutual_assured_destruction Why shouldn't the ship be contaminated by whatever weapon was used?

**Comment by Nechochwen-D** [comment]
@DingusShingleton as it stands effectiveness of everything but "radiation" goes down the greater the ratio of bunks:crew.

**Comment by Nechochwen-D** [comment]
@BorkBorkGoesTheCode the prospective damage types would not affect their parent ship unless in the form of a blast radius. In the case of "hull breaching" the projectile would be designed to "breach" the hull and expose the crew to vacuum. "Toxin" would work over time and linger providing the attacked ship with a bonus to defense when boarded.

**Comment by KiLEdEnNis** [comment]
isn't it possible that instead of killing crew, you could temporarily incapacitate a certain amount (to temporarily have an enemy ship lack crew and more eratically), but if you board them they can still all defend?
it would balance it and still give the desired effect

**Comment by MessyMix** [comment]
@KiLEdEnNis Good luck on writing the source. Not with you there. ;)

**Comment by KiLEdEnNis** [comment]
I can't do source, & the basic concept about disabling crew isn't my idea so... good luck Necho \( ͡° ͜ʖ ͡°)/

**Comment by SonEasterZombie** [comment]
perhaps the weapons could only kill crew when the ship is disabled, which would mean that any attempts to use it could easily result in the ships destruction, especially if the weapon did good hull damage and wasn't beam based. there could also be an alternate version of the weapon that would work like @KiLEdEnNis wanted

**Comment by KiLEdEnNis** [comment]
honestly I'd not go with any form of crew KILLING w/o boarding, incapacitating at most

**Comment by Nechochwen-D** [comment]
Incorporated feedback from the thread.

Let me know what you think.

**Comment by KiLEdEnNis** [comment]
I'd still think only incapacitating crew is better to keep boarding from becoming OP (as you'd kill defenders before boarding)
but the rest seems fine

**Comment by Nechochwen-D** [comment]
Incapacitating without the possibility of killing. Makes about as much sense as pokemon only ever fainting.
http://www.thegamercat.com/comic/chain-reaction/#comment-588725

**Comment by KiLEdEnNis** [comment]
but pokemon only faint which for some reason makes sense, considering you can fall from 10km height, get struck w/ lightning & get set on fire & surviving as you are the main character
but that's not even remotely on topic, cya

**Comment by johndh** [comment]
As far as balancing this weaponry, perhaps the player just isn't able to get it somehow?  All the ships that have it are set to self-destruct, for example, or possession of it is considered such an atrocity that you become persona non grata everywhere except the pirate worlds.

**Comment by Nechochwen-D** [comment]
If you look at the math the only _really_ dangerous damage type is "toxin" and it requires a successful hull breach to inject its' payload and I would think that they would only be installed on marauder vessels with neither the ammo or launchers available anywhere. "Hull breaching" I see as a natural side effect of having high power kinetic weapons impacting the hull. "Radiation" wouldn't be used all that much and is the least effective when the ship is in good shape but becomes a much greater threat once the shields have been dropped, it would be an attribute on the nukes and on some beam weapons.

**Comment by Tadrix** [comment]
Erratic movement is a crutch to punish players for boarding other ships too much, as a balancing tool, and that's the sole purpose it exists in current form.  Emergence of mechanics that build upon it is scary.

Radiation and toxin wouldn't do anything useful, really. In larger ships you have compartments separated by bulkheads and self-sealing emergency airlocks, so any such "damage" would stay where it is. It's also important to remember that ship's crew isn't equally scattered across the ship, but tend to be in specific places (like the bridge), and inability to target a specific place within the ship means there would need to be a random element (did manage to pierce a crewed compartment or not) before incapacitating crew.

We already have ionic damage (something akin to EMP) to do things similar as to what's described here — it makes a ship defenseless and temporarily "stunned" despite shields and what not.

**Comment by Bladewood** [comment]
On the ion damage issue, we also have heat damage.
Might as well give more variety.
We can even get the whole RGB thing. (red corresponds to heat, green corresponds to toxins and radiation, blue corresponds to ion)

**Comment by tehhowch** [comment]
As a counter to "kill crew then board at will": 
- a ship's tendency to self-destruct upon opening the airlock doors could be related to the fraction of crew lost. Did you kill every required crew member? Congrats, this ship might as well belong to automata. Only killed 10%? 10% more likely to \*poof\* if you decide you like the hull and not just the outfits. You'd always need to consider if you should use them, based on your goal with the target. Plundering? Fire-at-will. Capping? Maybe just 2-3 hits to help disable. If I use too many, they're likely to scuttle instead of allow boarding.
- Random self-destruct mechanisms would have to require that the shields first be depleted before the factors causing them can arise, preventing situations with lucky rolls giving instantly exploding ships.
    - random chance to self-destruct e.g. piercing the reactor/fuel cell. During the damage application, get a random number and compare to the ratio of `(<powerplant size> / <total outfit space>)^(some int)` (for example): if your random number is lower than that ratio, kaboom. 
    - An alternate source of explosion from `radioactive` weapons could be an accidental critical geometry created by successive piercing rounds. This would only be sensible for weapons using fissile ammo (U or Pu, really), and one formula for the probability of explosion could be `(numSuccessfulHits-1)^k * projectileMass / targetMass`, where `k` is chosen based on the projectile's crew lethality / damage, such that the ship will have exploded before you have killed all the crew. 

These side effects would ensure piercing / crew-killing weapons are firmly in the Secondary Weapons category.

Without the ability to hire bonus crew on your escorts, these outfits would be an absolute PITA to play against.

**Comment by Nechochwen-D** [comment]
@Tadrix 
>ship's crew isn't equally scattered across the ship, but tend to be in specific places

I can put in a system to account for this but without a broader spectrum of crew types I decided not to include it at this point.

>Consider player being the only ship member (any starting ship, for example). This would be unfair insta-kill.

According to the equations used the odds of that happening in a Sparrow with default stats for "hull breaching" (likely the most common attribute to be encountered in the early game) are 1:24 for kill and 1:4 for disable per point of "hull breaching". There would never be injury from toxin unless the total non dispersed toxin is greater than the filtration value. And radiation would be mostly negligible unless it is a massive value.
I will likely need to implement a "medical" system that works to heal "crewDisabled" even when the ship itself is disabled

> Consider a fleet with erratically moving escorts. If one is low on crew on the flagship, when time comes to land somewhere most likely there will be manual controls involved, because AI tends to ignore the need for much longer braking distance, see #2259 . Now if this is happening to escorts, the player can't do anything about it, no way to guide those ships, no way to replenish crew, no way to tow them along. It will take ages for escorts to be ready to land (and there's always the possibility that more ships equipped with "crew killers" will jump in). The result? Ensuing frustration and game reloading. This is gameplay breaking.

All that is needed is for escorts to be in the same system as the flag ship for them to be re-crewed when taking off.

@tehhowch 
I like the idea of increasing the odds of the ship self destructing in proportion to the number of crew killed However I do not really see the point of destroying outfits or causing spontaneous destruction.

I would also like to see a bit more definition in the whole "pilotError"system although having a certain number / percentage of crew that are hired above the required amount would be nice.

**Comment by Nechochwen-D** [comment]
I would also like to reinstate DeathBenifits as part of this system as well in a more realistic manner but I am not entirely sure if it would be a good idea at this point.

**Pull request closed without merging**